tests/test_logger.py

```diff
@@ -1,112 +1,130 @@
-        self.assertNotEqual(handler.buf[0]['line'], lineTwo)
-        logger.removeHandler(handler)
+    def handler_test(self):
+        handler = LogDNAHandler(LOGDNA_API_KEY, sample_options)
+        self.assertIsInstance(handler, logging.Handler)
+        self.assertIsInstance(handler.internal_handler, logging.StreamHandler)
+        self.assertIsNotNone(handler.internalLogger)
+        self.assertEqual(handler.key, LOGDNA_API_KEY)
+        self.assertEqual(handler.hostname, sample_options['hostname'])
+        self.assertEqual(handler.ip, sample_options['ip'])
+        self.assertEqual(handler.mac, sample_options['mac'])
+        self.assertEqual(handler.loglevel, 'info')
+        self.assertEqual(handler.app, '')
+        self.assertEqual(handler.env, '')
+        self.assertEqual(handler.tags, sample_options['tags'].split(','))
+        self.assertEqual(handler.custom_fields, defaults['META_FIELDS'])
+
+        # Set the Connection Variables
+        self.assertEqual(handler.url, defaults['LOGDNA_URL'])
+        self.assertEqual(handler.request_timeout,
+                         defaults['DEFAULT_REQUEST_TIMEOUT'])
+        self.assertEqual(handler.user_agent, defaults['USER_AGENT'])
+        self.assertEqual(handler.max_retry_attempts,
+                         defaults['MAX_RETRY_ATTEMPTS'])
+        self.assertEqual(handler.max_retry_jitter,
+                         defaults['MAX_RETRY_JITTER'])
+        self.assertEqual(handler.max_concurrent_requests,
+                         defaults['MAX_CONCURRENT_REQUESTS'])
+        self.assertEqual(handler.retry_interval_secs,
+                         defaults['RETRY_INTERVAL_SECS'])
+
+        # Set the Flush-related Variables
+        self.assertEqual(handler.buf, [])
+        self.assertEqual(handler.buf_size, 0)
+        self.assertEqual(handler.secondary, [])
+        self.assertFalse(handler.exception_flag)
+        self.assertIsNone(handler.flusher)
+        self.assertTrue(handler.index_meta)
+        self.assertEqual(handler.flush_limit, defaults['FLUSH_LIMIT'])
+        self.assertEqual(handler.flush_interval_secs,
+                         defaults['FLUSH_INTERVAL_SECS'])
+        self.assertEqual(handler.buf_retention_limit,
+                         defaults['BUF_RETENTION_LIMIT'])
+
+        # Set up the Thread Pools
+        self.assertIsInstance(handler.worker_thread_pool, ThreadPoolExecutor)
+        self.assertIsInstance(handler.request_thread_pool, ThreadPoolExecutor)
+        self.assertEqual(handler.level, logging.DEBUG)
+
+    def flusher_test(self):
+        handler = LogDNAHandler(LOGDNA_API_KEY, sample_options)
+        self.assertIsNone(handler.flusher)
+        handler.start_flusher()
+        self.assertIsNotNone(handler.flusher)
+        handler.close_flusher()
+        self.assertIsNone(handler.flusher)
+
+    def emit_test(self):
+        handler = LogDNAHandler(LOGDNA_API_KEY, sample_options)
+        handler.buffer_log = unittest.mock.Mock()
```

Comment:
This is meat of functionality, we shouldn't mock this. This leaves the coverage of the logger itself at ~50%

Reply:
That's correct! `buffer_log` is the core of the library but, @esatterwhite, in this case, what would your suggestion be in terms of testing `emit`? CC: @teeohhem

Reply:
Personally, if we're going to forgo the integration tests for now, I'd look into implementing a mock `ThreadPoolExecutor` that returns synchronously.

e.g.

```
class MockThreadPoolExecutor():
    def __init__(self, **kwargs):
        pass

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, exc_traceback):
        pass

    def submit(self, fn, *args, **kwargs):
        # execute functions in series without creating threads
        # for easier unit testing
        result = fn(*args, **kwargs)
        return result

    def shutdown(self, wait=True):
        pass
```

Then in the unit test you can do something like this:
```
with patch('logdna.ThreadPoolExecutor', wraps= MockThreadPoolExecutor()) as mock:
    #write tests against function and assert buffer contains x

Reply:
With the help of `MockThreadPoolExecutor`, thanks to @teeohhem, it's now 75+%